spring-boot/proxy/src/main/java/eu/coatrack/proxy/security/LocalApiKeyAndServiceApiManager.java

```diff
@@ -0,0 +1,171 @@
+ * calls which work even if the CoatRack admin server is down.
+ *
+ * @author Christoph Baier
+ */
+
+@EnableAsync
+@Service("localApiKeyAndServiceApiManager")
+public class LocalApiKeyAndServiceApiManager {
+
+    private static final Logger log = LoggerFactory.getLogger(LocalApiKeyAndServiceApiManager.class);
+    private final int fiveMinutesInMillis = 1000 * 60 * 5;
+
+    protected List<ApiKey> localApiKeyList = new ArrayList<>();
+    protected LocalDateTime latestLocalApiKeyListUpdate = LocalDateTime.now();
+
+    @Value("${minutes-the-gateway-works-without-connection-to-admin}")
+    private long minutesTheGatewayWorksWithoutConnectionToAdmin = 60;
+
+    private AdminCommunicator adminCommunicator;
+
+    public LocalApiKeyAndServiceApiManager(AdminCommunicator adminCommunicator) {
+        this.adminCommunicator = adminCommunicator;
+    }
+
+    public boolean isApiKeyValidConsideringTheLocalApiKeyList(String apiKeyValue) {
+        log.debug("Begin checking if the API key with the value {} is valid using the local API key list.",
+                apiKeyValue);
+        if (apiKeyValue == null) {
+            log.info("The passed API key value is null and can therefore not be checked for validity. " +
+                    "It is therefore rejected.");
+            return false;
+        }
+
+        ApiKey apiKey = findApiKeyFromLocalApiKeyList(apiKeyValue);
+        if (apiKey == null) {
+            return false;
+        } else
+            log.debug("The API key with the value {} is within the local API key list.", apiKeyValue);
+
+        return isApiKeyValid(apiKey);
+    }
+
+    private ApiKey findApiKeyFromLocalApiKeyList(String apiKeyValue) {
+        log.debug("Trying to find the service API associated to the API key with the value {} from the local list.",
+                apiKeyValue);
+        ApiKey apiKey;
+        try {
+            apiKey = localApiKeyList.stream().filter(apiKeyFromLocalList -> apiKeyFromLocalList.getKeyValue()
+                    .equals(apiKeyValue)).findFirst().get();
+        } catch (Exception e) {
+            log.info("The API key with the value {} can not be found within the local API key list " +
+                    "and is therefore rejected.", apiKeyValue);
+            return null;
+        }
+        return apiKey;
+    }
+
+    private boolean isApiKeyValid(ApiKey apiKey) {
+        boolean isApiKeyValid =
+                isApiKeyNotDeleted(apiKey) &&
+                        isApiKeyNotExpired(apiKey) &&
+                        wasLatestUpdateOfLocalApiKeyListWithinTheGivenDeadline();
+        if (isApiKeyValid)
+            log.info("The API key with the value {} is valid considering the local API key list and is " +
+                    "therefore accepted.", apiKey.getKeyValue());
+        return isApiKeyValid;
+    }
+
+    private boolean isApiKeyNotDeleted(ApiKey apiKey) {
+        boolean isNotDeleted = apiKey.getDeletedWhen() == null;
```

Comment:
The benefit of writing the method this way lies a few lines below that code snippet. If a key is deleted, a log info is created mentioning that the key is not valid because it is deleted. The one who is running the gateway exactly knows why a key was rejected. Imo this benefit outweighs the extra lines of code.

https://github.com/ChristophBaierATB/coatrack/blob/7e213271745f535c169ed3241ced4a8ca603edf5/spring-boot/proxy/src/main/java/eu/coatrack/proxy/security/LocalApiKeyAndServiceApiManager.java#L112-L120

Reply:
I the meantime I changed my mind about that topic. I think my logging efforts were too verbose and I shortened the code as you suggested.